### PR TITLE
Feature/1415 channel lifecycle status

### LINKED ISSUE #1415: Add support to get channel lifecycle status
Add support to get channel lifecycle status 

Channel lifecycle status is described at [Channel status / REST / [Docs]|[https://ably.com/docs/rest/channel-status#metadata-rest](https://ably.com/docs/rest/channel-status#metadata-rest)]

It is currently possible to retrieve channel lifecycle status in ably-go by using REST.Request and marshalling the result into a slice of empty interface.

**Example:**

```go
result, _ := restClient.Request("get", "/channels/test").Pages(context.Background())

result.Next(ctx)

var items []interface{}
if err := result.Items(&items); err != nil {
	fmt.Println(err)
}

fmt.Printf("%+v\n", items)
// [map[channelId:test name:test status:map[isActive:true occupancy:map[metrics:map[connections:1 presenceConnections:1 presenceMembers:1 presenceSubscribers:1 publishers:1 subscribers:1]]]]]
```

**Out of Scope**
Realtime subscription to Channel Lifecycle Events. This ticket is only for supporting REST client ‘Get’.

Please see this [https://github.com/ably/docs/pull/1415](https://github.com/ably/docs/pull/1415|smart-link)for more details

┆Issue is synchronized with this [Jira Story](https://ably.atlassian.net/browse/SDK-1848) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by maratal**
Closes #1415

### CODE REVIEW

**Review by lukasz-szyszkowski — APPROVED** [review]

File: Source/ARTRestChannel.m [hunk 1]
```diff
@@ -214,6 +218,57 @@ - (BOOL)history:(ARTDataQuery *)query callback:(ARTPaginatedMessagesCallback)cal
...
+        NSURL *url = [NSURL URLWithString:self->_basePath];
+        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
+        
+        [self.logger debug:__FILE__ line:__LINE__ message:@"RS:%p C:%p (%@) channel details request %@", self->_rest, self, self.name, request];
+        
+        [self->_rest executeRequest:request withAuthOption:ARTAuthenticationOn completion:^(NSHTTPURLResponse * _Nullable response, NSData * _Nullable data, NSError * _Nullable error) {
+            
+            if (response.statusCode == 200 /*OK*/) {
+                NSError *decodeError = nil;
+                ARTChannelDetails *channelDetails = [[self->_rest defaultEncoder] decodeChannelDetails:data error:&decodeError];
```
Comment (lawrence-forooghian) [inline Source/ARTRestChannel.m:240]: Instead of using `defaultEncoder`, should we use the MIME type of the response to make sure we use the correct encoder depending on whether we receive JSON or MessagePack in the response? See [this example of `-[ARTRest time:]`](https://github.com/ably/ably-cocoa/blob/0fe3c59658ce7d0b7181571cdad1f068d4085088/Source/ARTRest.m#L522).

File: Spec/Tests/RestClientChannelTests.swift [hunk 2]
```diff
@@ -1488,4 +1488,26 @@ class RestClientChannelTests: XCTestCase {
...
+    // ??
+    func test__047__lifecycle_status_response_should_contain_status_object() {
+        let channelName = uniqueChannelName()
+        let options = AblyTests.commonAppSetup()
+        let testToken = getTestToken(capability: "{\"*\":[\"channel-metadata\"]}")
+        options.token = testToken
+        let client = ARTRest(options: options)
+        let channel = client.channels.get(channelName)
+        
+        waitUntil(timeout: testTimeout) { done in
```
Comment (lawrence-forooghian) [inline Spec/Tests/RestClientChannelTests.swift:1501]: Can we test the values inside the `details` object? To do so might require a better understanding of what the values inside the `ARTChannelDetails` object mean (e.g. how do we make `presenceConnections` be a non-zero value etc) which I think will mean looking into the documentation for the API.

Also we might want to test the cases where `status` and `occupancy` (are / are not) null.

File: Spec/Tests/RestClientChannelTests.swift [hunk 3]
```diff
@@ -1488,4 +1488,26 @@ class RestClientChannelTests: XCTestCase {
             }
         }
     }
+    
+    // ??
```
Comment (lawrence-forooghian) [inline Spec/Tests/RestClientChannelTests.swift:1492]: Can we write a feature spec point number here?

File: Source/ARTTypes.m [hunk 4]
```diff
@@ -134,6 +134,78 @@ - (NSString *)description {
...
+
+- (instancetype)initWithChannelId:(NSString *)channelId status:(ARTChannelStatus *)status {
+    if (self = [super init]) {
+        _channelId = channelId;
+        _status = status;
+    }
+    return self;
+}
+
+- (instancetype)initWithChannelId:(NSString *)channelId status:(BOOL)status metrics:(ARTChannelMetrics *)metrics {
```
[the remark this answers is not in the document]
Comment (maratal) [reply Source/ARTTypes.m]: yep, this func no longer exists https://github.com/ably/ably-cocoa/pull/1420/commits/a68c3a0af033951c2069c3a38fddd9310a5d0db9

File: Source/ARTTypes.m [hunk 5]
```diff
@@ -134,6 +134,78 @@ - (NSString *)description {
...
+    ARTChannelOccupancy *occupancy = metrics != nil ? [[ARTChannelOccupancy alloc] initWithMetrics:metrics] : nil;
+    ARTChannelStatus *channelStatus = [[ARTChannelStatus alloc] initWithOccupancy:occupancy active:status];
+    ARTChannelDetails *details = [[ARTChannelDetails alloc] initWithChannelId:channelId status:channelStatus];
```
[the remark this answers is not in the document]
Comment (maratal) [reply Source/ARTTypes.m]: duh.. my bad https://github.com/ably/ably-cocoa/pull/1420/commits/a68c3a0af033951c2069c3a38fddd9310a5d0db9

File: Source/ARTJsonLikeEncoder.m [hunk 6]
```diff
@@ -753,6 +757,28 @@ - (ARTConnectionDetails *)connectionDetailsFromDictionary:(NSDictionary *)input
                                          maxIdleInterval:millisecondsToTimeInterval([input artInteger:@"maxIdleInterval"])];
 }
 
+- (ARTChannelDetails *)channelDetailsFromDictionary:(NSDictionary *)input {
+    if (!input) {
+        return nil;
+    }
+    
+    NSDictionary* statusDict = [input valueForKey:@"status"];
+    NSDictionary* metricsDict = [statusDict valueForKeyPath:@"occupancy.metrics"];
```
[the remark this answers is not in the document]
Comment (maratal) [reply Source/ARTJsonLikeEncoder.m]: https://github.com/ably/ably-cocoa/pull/1420/commits/a68c3a0af033951c2069c3a38fddd9310a5d0db9

File: Source/ARTTypes.h [hunk 7]
```diff
@@ -201,6 +201,62 @@ NSString *generateNonce(void);
...
+
+@end
+
+#pragma mark - ARTChannelDetails
+
+@interface ARTChannelDetails : NSObject
+
+@property (nonatomic, strong, readonly) NSString *channelId;
+
+@property (nonatomic, strong, readonly) ARTChannelStatus *status;
```
[the remark this answers is not in the document]
Comment (maratal) [reply Source/ARTTypes.h]: now nullable in code https://github.com/ably/ably-cocoa/pull/1420/commits/a68c3a0af033951c2069c3a38fddd9310a5d0db9

File: Source/ARTTypes.m [hunk 8]
```diff
@@ -134,6 +134,78 @@ - (NSString *)description {
...
+    if (self = [super init]) {
+        _channelId = channelId;
+        _status = status;
+    }
+    return self;
+}
+
+- (instancetype)initWithChannelId:(NSString *)channelId status:(BOOL)status metrics:(ARTChannelMetrics *)metrics {
+    ARTChannelOccupancy *occupancy = metrics != nil ? [[ARTChannelOccupancy alloc] initWithMetrics:metrics] : nil;
+    ARTChannelStatus *channelStatus = [[ARTChannelStatus alloc] initWithOccupancy:occupancy active:status];
```
[the remark this answers is not in the document]
Comment (maratal) [reply Source/ARTTypes.m]: https://github.com/ably/ably-cocoa/pull/1420/commits/a68c3a0af033951c2069c3a38fddd9310a5d0db9

File: Source/ARTRestChannel.m (hunk 1 above)
Comment (maratal) [reply Source/ARTRestChannel.m]: https://github.com/ably/ably-cocoa/pull/1420/commits/f83d845dd5f51e680acec4a04a323eabbeb356a0

File: Spec/Tests/RestClientChannelTests.swift (hunk 3 above)
Comment (maratal) [reply Spec/Tests/RestClientChannelTests.swift]: https://github.com/ably/ably-cocoa/pull/1420/commits/8d9917c10bb226641e00e3176351519749a36940

File: Source/ARTTypes.h [hunk 9]
```diff
@@ -201,6 +201,62 @@ NSString *generateNonce(void);
...
+
+@end
+
+#pragma mark - ARTChannelStatus
+
+@interface ARTChannelStatus : NSObject
+
+@property (nonatomic, readonly) BOOL active;
+
+@property (nonatomic, strong, readonly, nullable) ARTChannelOccupancy *occupancy;
```
[the remark this answers is not in the document]
Comment (maratal) [reply Source/ARTTypes.h:240]: https://github.com/ably/docs/pull/1455

File: Source/ARTTypes.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (maratal) [reply Source/ARTTypes.h]: Spec PR - https://github.com/ably/docs/pull/1455

File: Spec/Tests/RestClientChannelTests.swift (hunk 2 above)
Comment (maratal) [reply Spec/Tests/RestClientChannelTests.swift:1517]: As [this comment](https://github.com/ably/docs/pull/1455#issuecomment-1153664273) says, these fields should never be nil.
https://github.com/ably/ably-cocoa/pull/1420/commits/8c4610825ced2c1c6d51c19879ea9f46d1ee977e
Test with counters check: https://github.com/ably/ably-cocoa/pull/1420/commits/ab35c8bf3a2931fdbee4f049d695b10ba33b70fb

File: Source/ARTRestChannel.m (hunk 1 above)
Comment (lawrence-forooghian) [reply Source/ARTRestChannel.m]: Cool, I've just noticed we should probably also set the `Accept` header on the request, [again similarly to `-[ARTRest time:]`](https://github.com/ably/ably-cocoa/blob/0fe3c59658ce7d0b7181571cdad1f068d4085088/Source/ARTRest.m#L509-L510).

Also, I just noticed another issue with `-[ARTRest time:]` which I think applies here, too: https://github.com/ably/ably-cocoa/issues/1442. If this is the case, please could you make an issue (or update the existing one) so that we address it here too?

File: Spec/Tests/RestClientChannelTests.swift [hunk 10]
```diff
@@ -1488,4 +1488,47 @@ class RestClientChannelTests: XCTestCase {
...
+            realtimeChannel.subscribe { message in
+                done()
+            }
+            restChannel.publish(nil, data: nil) { error in
+                expect(error).to(beNil())
+            }
+        }
+        
+        waitUntil(timeout: testTimeout) { done in
+            restChannel.status { details, error in
```
[the remark this answers is not in the document]
Comment (SimonWoolf) [reply Spec/Tests/RestClientChannelTests.swift:1518]: It's a currently-pointless property that's always true

File: Spec/Tests/RestClientChannelTests.swift [hunk 11]
```diff
@@ -1488,4 +1488,47 @@ class RestClientChannelTests: XCTestCase {
             }
         }
     }
+    
+    // RSL8a, CHD2b, CHS2b, CHO2a
+    func test__047__channel_with_subscribers_status_method_returns_a_channel_details_object_with_fullfilled_channel_metrics() {
```
[the remark this answers is not in the document]
Comment (maratal) [reply Spec/Tests/RestClientChannelTests.swift]: https://github.com/ably/ably-cocoa/pull/1420/commits/31503e8339246b3e0db96209a43cb28a1930dffe

File: Spec/Tests/RestClientChannelTests.swift [hunk 12]
```diff
@@ -1488,4 +1488,47 @@ class RestClientChannelTests: XCTestCase {
...
+            restChannel.status { details, error in
+                expect(error).to(beNil())
+                guard let details = details else {
+                    fail("Channel details are empty"); done()
+                    return
+                }
+                expect(details.status.occupancy.metrics.connections) == 1 // CHM2a
+                expect(details.status.occupancy.metrics.publishers) == 1 // CHM2e
+                expect(details.status.occupancy.metrics.subscribers) == 1 // CHM2f
+                expect(details.status.occupancy.metrics.presenceMembers) == 0 // CHM2c
```
[the remark this answers is not in the document]
Comment (maratal) [reply Spec/Tests/RestClientChannelTests.swift]: https://github.com/ably/ably-cocoa/pull/1420/commits/20ded1dc68855cae2e09e13ce21e5a93f8246826

File: Source/ARTRestChannel.m [hunk 13]
```diff
@@ -214,6 +218,58 @@ - (BOOL)history:(ARTDataQuery *)query callback:(ARTPaginatedMessagesCallback)cal
...
+        ARTChannelDetailsCallback userCallback = callback;
+        callback = ^(ARTChannelDetails *details, ARTErrorInfo *_Nullable error) {
+            dispatch_async(self->_userQueue, ^{
+                userCallback(details, error);
+            });
+        };
+    }
+    dispatch_async(_queue, ^{
+        NSURL *url = [NSURL URLWithString:self->_basePath];
+        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
```
[the remark this answers is not in the document]
Comment (maratal) [reply Source/ARTRestChannel.m:232]: >I've just noticed we should probably also set the Accept header on the request, [again similarly to -[ARTRest time:]]

I think it's a legacy code. `Accept` header is set later here for all requests:
https://github.com/ably/ably-cocoa/blob/0fe3c59658ce7d0b7181571cdad1f068d4085088/Source/ARTRest.m#L313

[the remark this answers is not in the document]
Comment (maratal) [reply Source/ARTRestChannel.m:232]: >If this is the case, please could you make an issue (or update the existing one) so that we address it here too?

I've replied there https://github.com/ably/ably-cocoa/issues/1442#issuecomment-1179390433

[the remark this answers is not in the document]
Comment (maratal) [reply Source/ARTRestChannel.m:232]: But there is still an issue with this
```
id<ARTEncoder> encoder = self->_rest.encoders[response.MIMEType];
ARTChannelDetails *channelDetails = [encoder decodeChannelDetails:data error:&decodeError];
```
`channelDetails` can be nil if `encoder` is nil and the callback will be called with both params being `nil`.

File: Spec/Tests/RestClientChannelTests.swift [hunk 14]
```diff
@@ -1488,4 +1488,47 @@ class RestClientChannelTests: XCTestCase {
...
+    
+    // RSL8a, CHD2b, CHS2b, CHO2a
+    func test__047__channel_with_subscribers_status_method_returns_a_channel_details_object_with_fullfilled_channel_metrics() {
+        let options = AblyTests.commonAppSetup()
+        let rest = ARTRest(options: options)
+        let realtime = ARTRealtime(options: options)
+
+        let channelName = uniqueChannelName()
+        
+        let realtimeChannel = realtime.channels.get(channelName)
```
[the remark this answers is not in the document]
Comment (maratal) [reply Spec/Tests/RestClientChannelTests.swift:1499]: In addition to the new descriptions, turned out that to increase _all_ `metrics` by 1 it's enough to call `enter:` in `presence` in realtime channel - https://github.com/ably/ably-cocoa/pull/1420/commits/a1b14095b3822be654a08d04818e5eeaa8f30c40

File: Source/ARTJsonLikeEncoder.m [hunk 15]
```diff
@@ -765,26 +765,23 @@ - (ARTChannelMetrics *)channelMetricsFromDictionary:(NSDictionary *)input {
                                       presenceSubscribers:[input artInteger:@"presenceSubscribers"]];;
 }
 
-- (nullable ARTChannelOccupancy *)channelOccupancyFromDictionary:(NSDictionary *)input {
-    NSDictionary *metricsDict = [input valueForKey:@"metrics"];
-    if (metricsDict == nil) {
-        return nil;
-    }
+- (ARTChannelOccupancy *)channelOccupancyFromDictionary:(NSDictionary *)input {
+    NSDictionary *metricsDict = [input valueForKey:@"metrics"] ?: @{};
```
[the remark this answers is not in the document]
Comment (maratal) [reply Source/ARTJsonLikeEncoder.m]: https://github.com/ably/ably-cocoa/pull/1420/commits/4777c543399f0baa2e377d4c37afb0b5f7c342a5

File: Spec/Tests/RestClientChannelTests.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (maratal) [reply Spec/Tests/RestClientChannelTests.swift:1499]: @SimonWoolf Couldn't you please reassure that this is expected behavior for all metrics in this case? Also I would appreciate if you suggest a proper test name for it, f.e. `test__047__channel_with_presence_members_status_method_returns_a_channel_details_object_with_fullfilled_channel_metrics`, does it accurately reflect what's going on here? Thanks

[the remark this answers is not in the document]
Comment (SimonWoolf) [reply Spec/Tests/RestClientChannelTests.swift:1499]: > please reassure that this is expected behavior for all metrics in this case?

yes

tbh I'm not sure I'd worry about all of this that much? In the sense that: if the client library is successfully obtaining the occupancy object, it has done all that it can or should do; making sure that the returned object is _accurate_ is the server's responsibility

[the remark this answers is not in the document]
Comment (lawrence-forooghian) [reply Spec/Tests/RestClientChannelTests.swift:1499]: > tbh I'm not sure I'd worry about all of this that much? In the sense that: if the client library is successfully obtaining the occupancy object, it has done all that it can or should do; making sure that the returned object is accurate is the server's responsibility

Agreed, but since our general approach to tests in this SDK has been one that uses very little mocking, we don't really have a way to verify that the SDK is successfully parsing the object contained in the response other than by creating a scenario that provokes a certain expected response from the server and then asserting the object returned by the SDK matches that expected response. It's probably not ideal.

**Review by lawrence-forooghian — APPROVED** [review]
